runtimes/go/storage/objects/internal/providers/gcs/bucket.go

```diff
@@ -178,6 +190,55 @@ func (b *bucket) Attrs(data types.AttrsData) (*types.ObjectAttrs, error) {
 	return mapAttrs(resp), mapErr(err)
 }
 
+func (b *bucket) SignedUploadURL(data types.UploadURLData) (string, error) {
+	// GCS allows a max of seven days TTL. We catch that here to return the
+	// right error type. The GCS client returns a generic error for this that
+	// is hard to distinguish from others.
+	if data.Ttl > 7*24*time.Hour {
+		return "", types.ErrPreconditionFailed
+		// TODO: would be nice to return a specific error message with this one.
+	}
+
+	opts := &storage.SignedURLOptions{
+		Scheme:  storage.SigningSchemeV4,
+		Method:  "PUT",
+		Expires: time.Now().Add(time.Duration(data.Ttl)),
+	}
+
+	// We use a fake GCS service for local development. Ideally, the runtime
+	// code would be oblivious to this once the GCS client is set up. But that
+	// turns out to be difficult for URL signing, so we add a special case
+	// here.
+	if b.localSign != nil {
+		opts.GoogleAccessID = b.localSign.accessId
+		opts.PrivateKey = []byte(b.localSign.privateKey)
+	}
+
+	url, err := b.handle.SignedURL(data.Object.String(), opts)
+	if err != nil {
+		return "", mapErr(err)
+	}
+
+	// More special handling for the local dev case.
+	if b.localSign != nil {
+		url = replaceURLPrefix(url, b.localSign.baseUrl)
+	}
+
+	return url, nil
+}
+
+func replaceURLPrefix(orig_url string, base string) string {
+	u, err := url.Parse(orig_url)
+	if err != nil {
+		return orig_url // If the input URL is not valid, just return it as-is
+	}
+	out := strings.TrimRight(base, "/") + "/" + strings.TrimLeft(u.Path, "/")
```

Comment:
This adds an extra trailing slash when `u.Path` is empty, right?
```suggestion
	out := strings.TrimRight(base, "/")
	if u.Path != "" {
	    out += "/" + strings.TrimLeft(u.Path, "/")
	}
```

Reply:
Done (and `gcp/bucket.rs`)